engines/docker/sandbox.go

```diff
@@ -0,0 +1,156 @@
+package dockerengine
+
+import (
+	"context"
+	"time"
+
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/taskcluster/taskcluster-worker/engines"
+	"github.com/taskcluster/taskcluster-worker/runtime"
+	"github.com/taskcluster/taskcluster-worker/runtime/atomics"
+	"github.com/taskcluster/taskcluster-worker/runtime/caching"
+	"github.com/taskcluster/taskcluster-worker/runtime/ioext"
+)
+
+const dockerEngineKillTimeout = 5 * time.Second
+
+type sandbox struct {
+	engines.SandboxBase
+	monitor     runtime.Monitor
+	containerID string
+	resultSet   engines.ResultSet
+	resultErr   error
+	abortErr    error
+	tempStorage runtime.TemporaryStorage
+	resolve     atomics.Once
+	client      *docker.Client
+	taskCtx     *runtime.TaskContext
+	handle      *caching.Handle
+}
+
+func newSandbox(sb *sandboxBuilder) (*sandbox, error) {
+	// create the container
+	opts := docker.CreateContainerOptions{
+		Config: sb.generateDockerConfig(),
+	}
+
+	container, err := sb.e.client.CreateContainer(opts)
+	if err != nil {
+		return nil, runtime.NewMalformedPayloadError(
+			"could not create container: " + err.Error())
+	}
+
+	// create a temporary storage for use by resultSet
+	ts, err := sb.e.Environment.TemporaryStorage.NewFolder()
+	if err != nil {
+		// unsure if this is the correct error type to return
+		return nil, runtime.NewMalformedPayloadError(
+			"could not create temporary storage")
+	}
+	sbox := &sandbox{
+		containerID: container.ID,
+		tempStorage: ts,
+		client:      sb.e.client,
+		taskCtx:     sb.taskCtx,
+		handle:      sb.handle,
+		monitor:     sb.monitor.WithPrefix("docker-sandbox").WithTag("taskID", sb.taskCtx.TaskID),
+	}
+
+	// attach to the container before starting so that we get all the logs
+	_, err = sbox.client.AttachToContainerNonBlocking(docker.AttachToContainerOptions{
+		Container:    container.ID,
+		OutputStream: ioext.WriteNopCloser(sbox.taskCtx.LogDrain()),
+		Logs:         true,
+		Stdout:       true,
+		Stderr:       true,
+		Stream:       true,
+	})
+
+	// HostConfig is ignored by the remote API and is only kept for
+	// backward compatibility.
+	err = sbox.client.StartContainer(sbox.containerID, &docker.HostConfig{})
+	if err != nil {
+		return nil, runtime.ErrFatalInternalError
+	}
+
+	go sbox.wait()
+
+	return sbox, nil
+}
+
+func (s *sandbox) WaitForResult() (engines.ResultSet, error) {
+	s.resolve.Wait()
+	return s.resultSet, s.resultErr
+}
+
+func (s *sandbox) wait() {
+	exitCode, err := s.client.WaitContainer(s.containerID)
+	s.resolve.Do(func() {
+		s.resultSet = newResultSet(exitCode == 0, s.containerID, s.client,
+			s.tempStorage, s.handle, s.monitor.WithPrefix("result-set"))
+		if err != nil {
+			s.resultSet = newResultSet(false, s.containerID, s.client, s.tempStorage,
```

Comment:
Maybe don't create `resultSet` twice if `err != nil`

I suggest that you first determine if it's successful or not and then call `newResultSet`, like:
```go
success := exitCode == 0 && err == nil
s.resultSet = newResultSet(
  success,
  ...
)
```

use multiple lines if the call has a lot of parameters... Also give it `engine` instead of `s.tempStorage`, as we can access `environment.TemporaryStorage` from `engine`...